src/main/java/org/jabref/logic/texparser/TexBibEntriesResolver.java

```diff
@@ -1,61 +1,58 @@
+import org.jabref.model.util.FileUpdateMonitor;
 
 public class TexBibEntriesResolver {
 
     private final BibDatabase masterDatabase;
+    private final ImportFormatPreferences importFormatPreferences;
+    private final FileUpdateMonitor fileMonitor;
 
-    public TexBibEntriesResolver(BibDatabase masterDatabase) {
+    public TexBibEntriesResolver(BibDatabase masterDatabase, ImportFormatPreferences importFormatPreferences, FileUpdateMonitor fileMonitor) {
         this.masterDatabase = masterDatabase;
+        this.importFormatPreferences = importFormatPreferences;
+        this.fileMonitor = fileMonitor;
     }
 
     /**
-     * Look for BibTeX entries within the reference database for all keys inside of the TEX files.
-     * Insert these data in the list of new entries.
+     * Resolve all BibTeX entries and check if they are in the given database.
      */
-    public TexBibEntriesResolverResult resolveKeys(TexParserResult texParserResult) {
-        TexBibEntriesResolverResult result = new TexBibEntriesResolverResult(texParserResult);
-        Set<String> keySet = result.getCitationsKeySet();
-
-        for (String key : keySet) {
-            if (result.isNotKeyIntoNewEntries(key)) {
-                Optional<BibEntry> entry = masterDatabase.getEntryByKey(key);
-
-                if (entry.isPresent()) {
-                    result.insertEntry(entry.get());
-                    resolveCrossReferences(result, entry.get());
-                } else {
-                    result.addUnresolvedKey(key);
-                }
-            }
-        }
-
-        return result;
-    }
-
-    /**
-     * Find cross references for inserting into the list of new entries.
-     */
-    private void resolveCrossReferences(TexBibEntriesResolverResult result, BibEntry entry) {
-        entry.getField(StandardField.CROSSREF).ifPresent(crossRef -> {
-            if (result.isNotKeyIntoNewEntries(crossRef)) {
-                Optional<BibEntry> refEntry = masterDatabase.getEntryByKey(crossRef);
-
-                if (refEntry.isPresent()) {
-                    result.insertEntry(refEntry.get());
-                    result.increaseCrossRefsCount();
-                } else {
-                    result.addUnresolvedKey(crossRef);
-                }
-            }
-        });
+    public TexBibEntriesResolverResult resolve(TexParserResult texParserResult) {
+        TexBibEntriesResolverResult resolverResult = new TexBibEntriesResolverResult(texParserResult);
+
+        // Preload databases from BIB files.
+        Map<Path, BibDatabase> bibDatabases = resolverResult.getBibFiles().values().stream().distinct().collect(Collectors.toMap(
+                Function.identity(), path -> OpenDatabase.loadDatabase(path.toString(), importFormatPreferences, fileMonitor).getDatabase()));
+
+        // Add all new entries to the newEntries set.
+        resolverResult.getNewEntries().addAll(
+                // Get all pairs Entry<String entryKey, Citation>.
```

Comment:
please separate this to multiple statements, it's very hard to undersatnd otherwise.